unixfs/src/file/reader.rs

```diff
@@ -0,0 +1,333 @@
+    ) -> Result<FileReader<'a>, FileReadFailed> {
+        self.last_ending
+            .check_is_suitable_next(self.last_offset, &tree_range)?;
+        FileReader::from_continued(self, tree_range.start, next_block)
+    }
+}
+
+impl AsRef<FileMetadata> for Traversal {
+    fn as_ref(&self) -> &FileMetadata {
+        &self.metadata
+    }
+}
+
+/// Files in unixfs merkle trees can either contain content of the file, or can contain links to
+/// other parts of the tree.
+pub enum FileContent<'a, I>
+where
+    I: Iterator<Item = (PBLink<'a>, Range<u64>)> + 'a,
+{
+    /// When reaching the leaf level of a DAG we finally find the actual content. For empty files
+    /// without content this will be an empty slice.
+    Just(&'a [u8]),
+    /// The content of the file is spread over a number of blocks; iteration must follow from index
+    /// depth-first from the first link to reach the given the bytes in the given byte offset
+    /// range.
+    Spread(I),
+}
+
+impl<'a, I> FileContent<'a, I>
+where
+    I: Iterator<Item = (PBLink<'a>, Range<u64>)>,
+{
+    /// Returns the content as bytes, or panics if there were links instead.
+    pub fn unwrap_content(self) -> &'a [u8] {
+        match self {
+            FileContent::Just(x) => x,
+            y => panic!("Expected FileContent::Just, found: {:?}", y),
+        }
+    }
+}
+
+impl<'a, I> fmt::Debug for FileContent<'a, I>
+where
+    I: Iterator<Item = (PBLink<'a>, Range<u64>)>,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            FileContent::Just(bytes) => write!(fmt, "Just({} bytes)", bytes.len()),
+            FileContent::Spread(_) => write!(fmt, "Spread(...)"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Ending;
+    use crate::file::{FileError, FileReadFailed};
+
+    #[test]
+    fn collapsing_tree() {
+        // this is pretty much how I planned the ending might be useful but it's perhaps a bit
+        // confusing as it's only the half of the range
+        Ending::TreeCoverage(100)
+            .check_is_suitable_next(0, &(0..100))
+            .unwrap();
+        Ending::TreeCoverage(100)
+            .check_is_suitable_next(0, &(0..10))
+            .unwrap();
+        Ending::TreeCoverage(100)
+            .check_is_suitable_next(0, &(0..2))
+            .unwrap();
+        Ending::Chunk(2)
+            .check_is_suitable_next(0, &(2..10))
+            .unwrap();
+        Ending::TreeCoverage(10)
+            .check_is_suitable_next(2, &(2..10))
+            .unwrap();
+        Ending::TreeCoverage(10)
+            .check_is_suitable_next(2, &(10..20))
+            .unwrap();
+        Ending::Chunk(10)
+            .check_is_suitable_next(2, &(10..100))
+            .unwrap();
+    }
+
+    #[test]
+    fn expanding_tree() {
+        match Ending::TreeCoverage(100).check_is_suitable_next(10, &(0..102)) {
+            Err(FileReadFailed::File(FileError::TreeExpandsOnLinks)) => {}
```

Comment:
as before, I'd use `assert!(matches!(obj, pattern), "panic msg", panic args)` for clarity here and in the next few cases

Reply:
I did try to do this with macros as well but `$x:expr` cannot be put into `match { $x => { ... }}`.

Your suggested pattern would be:

```
let res = <expr>;
assert!(matches!(res, FileReadFailed::File(FileError::TreeExpandsOnLinks), "{:?}", res);
```

It is less lines at least before rustfmt and it'd seem like when handling the clippy errors I forgot to drop the `FileReadFailed` from here, perhaps I'll handle this first.

Reply:
Ah, fair, didn't notice it's an expression.

Reply:
Done.